client/asset/eth/eth.go

```diff
@@ -2371,6 +2462,122 @@ func (eth *ETHWallet) checkForNewBlocks(reportErr func(error)) {
 			w.checkFindRedemptions()
 		}
 	}()
+	go func() {
+		for _, w := range eth.walletList() {
+			w.checkConfirmRedemptions()
+		}
+	}()
+}
+
+func (w *assetWallet) swapIsRedeemed(secretHash [32]byte, contractVersion uint32) (bool, error) {
+	swap, err := w.swap(w.ctx, secretHash, contractVersion)
+	if err != nil {
+		return false, err
+	}
+	return swap.State == dexeth.SSRedeemed, nil
+}
+
+// confirmRedemption checks one of the queued redemption requests.
+func (w *assetWallet) confirmRedemption(txHash common.Hash, req *confirmRedemptionRequest, currentTip int64) {
+	const (
+		TxConfsNeededToConfirm               = 10
+		BlocksToWaitBeforeCoinNotFound       = 5
+		BlocksToWaitBeforeCheckingIfReplaced = 10
+	)
+
+	var blocksSinceRequestStart int64
+	if currentTip > req.startHeight {
+		blocksSinceRequestStart = currentTip - req.startHeight
+	}
+
+	tx, txBlock, err := w.node.getTransaction(w.ctx, txHash)
+	if err == asset.CoinNotFoundError {
+		w.log.Errorf("confirmRedemption for tx %s: could not find coin", txHash)
+		// Wait a few blocks before returning that the coin cannot be found.
+		if blocksSinceRequestStart < BlocksToWaitBeforeCoinNotFound {
+			return
+		}
+	}
+	if err != nil {
+		// If we cannot find a transaction, check if the swap has been redeemed
+		// by another transaction. If so, we can end the search.
+		redeemed, redeemedErr := w.swapIsRedeemed(req.secretHash, req.contractVersion)
+		if redeemedErr != nil {
+			// Just log this error, and return the error from getTransaction.
+			w.log.Errorf("unable to determine if swap has been redeemed: %v", err)
+		}
+		if redeemedErr != nil || !redeemed {
+			req.res <- err
+		} else {
+			// The swap was redeemed by another transaction.
+			req.res <- nil
+		}
+		return
+	}
+
+	var confirmations int64
+	if txBlock > 0 {
+		if txBlock > currentTip {
+			confirmations = 1
+		} else {
+			confirmations = currentTip - txBlock + 1
+		}
+	}
+
+	if confirmations >= TxConfsNeededToConfirm {
+		req.res <- nil
+		return
+	} else if confirmations > 0 {
+		return
+	}
+
+	// If the transaction is still unconfirmed, but was submitted less than 10
+	// blocks ago, give it some time before checking if it has been replaced,
+	// or redeemed by another tx.
+	if blocksSinceRequestStart < BlocksToWaitBeforeCheckingIfReplaced {
```

Comment:
Do you mean

```golang
if blocksSinceRequestStart%blocksToWaitBeforeCheckingIfReplaced != 0 || blocksSinceRequestStart == 0 {
```

since it's returning if this is true?

Reply:
Oh right.